### ISSUE TITLE
Crashes from clearPartsFromFilesystemImpl at 23.8

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#64797]
Sporadic crashes like that

```
2875989:2024.06.03 06:46:47.093275 [ 79410 ] {} <Fatal> BaseDaemon: ########## Short fault info ############
2875991:2024.06.03 06:46:47.093342 [ 79410 ] {} <Fatal> BaseDaemon: (version 23.8.6.16 (official build), build id: D093C714F52A7D6CE5C33DF3F8FED89E3C610AEE, git hash: 077df679bed122ad45c8b105d8916ccfec85ae64) (from thread 74523) Received signal 11
2875993:2024.06.03 06:46:47.093389 [ 79410 ] {} <Fatal> BaseDaemon: Signal description: Segmentation fault
2876008:2024.06.03 06:46:47.093408 [ 79410 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
2876012:2024.06.03 06:46:47.094003 [ 79410 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000000c625d14 0x0000000012c9ad00 0x0000000011073c1a 0x0000000010b362cf 0x000000000c7268be 0x000000000c72a37c 0x000000000c728bc4 0x00007ffff77ab1cf 0x00007ffff7417dd3
2876013:2024.06.03 06:46:47.094029 [ 79410 ] {} <Fatal> BaseDaemon: ########################################
2876014:2024.06.03 06:46:47.094048 [ 79410 ] {} <Fatal> BaseDaemon: (version 23.8.6.16 (official build), build id: D093C714F52A7D6CE5C33DF3F8FED89E3C610AEE, git hash: 077df679bed122ad45c8b105d8916ccfec85ae64) (from thread 74523) (no query) Received signal Segmentation fault (11)
2876015:2024.06.03 06:46:47.094065 [ 79410 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
2876016:2024.06.03 06:46:47.094068 [ 79410 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000000c625d14 0x0000000012c9ad00 0x0000000011073c1a 0x0000000010b362cf 0x000000000c7268be 0x000000000c72a37c 0x000000000c728bc4 0x00007ffff77ab1cf 0x00007ffff7417dd3
2876065:2024.06.03 06:46:47.095930 [ 79410 ] {} <Fatal> BaseDaemon: 2. std::pair<std::__hash_iterator<std::__hash_node<String, void*>*>, bool> std::__hash_table<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>::__emplace_unique_key_args<String, String const&>(String const&, String const&) @ 0x000000000c625d14 in /usr/bin/clickhouse
2876069:2024.06.03 06:46:47.096134 [ 79410 ] {} <Fatal> BaseDaemon: 3. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::MergeTreeData::clearPartsFromFilesystemImpl(std::vector<std::shared_ptr<DB::IMergeTreeDataPart const>, std::allocator<std::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::unordered_set<String, std::hash<String>, std::equal_to<String>, std::allocator<String>>*)::$_4, void ()>>(std::__function::__policy_storage const*) @ 0x0000000012c9ad00 in /usr/bin/clickhouse
2876094:2024.06.03 06:46:47.097455 [ 79410 ] {} <Fatal> BaseDaemon: 4. std::__packaged_task_func<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority)::'lambda'(), std::allocator<std::function<std::future<void> (std::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, String const&)::'lambda'(std::function<void ()>&&, Priority)::operator()(std::function<void ()>&&, Priority)::'lambda'()>, void ()>::operator()() @ 0x0000000011073c1a in /usr/bin/clickhouse
2876095:2024.06.03 06:46:47.097482 [ 79410 ] {} <Fatal> BaseDaemon: 5. std::packaged_task<void ()>::operator()() @ 0x0000000010b362cf in /usr/bin/clickhouse
2876097:2024.06.03 06:46:47.097496 [ 79410 ] {} <Fatal> BaseDaemon: 6. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000c7268be in /usr/bin/clickhouse
2876098:2024.06.03 06:46:47.097527 [ 79410 ] {} <Fatal> BaseDaemon: 7. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000c72a37c in /usr/bin/clickhouse
2876099:2024.06.03 06:46:47.097542 [ 79410 ] {} <Fatal> BaseDaemon: 8. void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000c728bc4 in /usr/bin/clickhouse
2876101:2024.06.03 06:46:47.097573 [ 79410 ] {} <Fatal> BaseDaemon: 9. start_thread @ 0x00000000000081cf in /usr/lib64/libpthread-2.28.so
2876102:2024.06.03 06:46:47.097601 [ 79410 ] {} <Fatal> BaseDaemon: 10. clone @ 0x0000000000039dd3 in /usr/lib64/libc-2.28.so
2894519:2024.06.03 06:46:51.885162 [ 79410 ] {} <Fatal> BaseDaemon: Integrity check of the executable successfully passed (checksum: 26AEDB409AB50F57AAF20243F5DEE734)
2894521:2024.06.03 06:46:51.885225 [ 79410 ] {} <Fatal> BaseDaemon: Report this error to https://github.com/ClickHouse/ClickHouse/issues
```

addr2line: 

```
llvm-addr2line-18 -afiCe /usr/lib/debug/usr/bin/clickhouse.debug 0x000000000c625d14 0x0000000012c9ad00 0x0000000011073c1a 0x0000000010b362cf 0x000000000c7268be 0x000000000c72a37c 0x000000000c728bc4 0x00007ffff77ab1cf 0x00007ffff7417dd3
0xc625d14
std::__1::pair<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, void*>*>, bool> std::__1::__hash_table<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>::__emplace_unique_key_args<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)
./build_docker/./contrib/llvm-project/libcxx/include/__hash_table:2000
0x12c9ad00
double std::__1::__cxx_atomic_load[abi:v15000]<double>(std::__1::__cxx_atomic_base_impl<double> const*, std::__1::memory_order)
./build_docker/./contrib/llvm-project/libcxx/include/atomic:958
std::__1::__atomic_base<double, false>::load[abi:v15000](std::__1::memory_order) const
./build_docker/./contrib/llvm-project/libcxx/include/atomic:1588
pthread_mutex_unlock
./build_docker/./src/Common/ThreadFuzzer.cpp:407
std::__1::__libcpp_mutex_unlock[abi:v15000](pthread_mutex_t*)
./build_docker/./contrib/llvm-project/libcxx/include/__threading_support:314
std::__1::mutex::unlock()
./build_docker/./contrib/llvm-project/libcxx/src/mutex.cpp:52
std::__1::lock_guard<std::__1::mutex>::~lock_guard[abi:v15000]()
./build_docker/./contrib/llvm-project/libcxx/include/__mutex_base:100
DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*)::$_4::operator()() const
./build_docker/./src/Storages/MergeTree/MergeTreeData.cpp:2471
decltype(std::declval<DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*)::$_4&>()()) std::__1::__invoke[abi:v15000]<DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*)::$_4&>(DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*)::$_4&)
./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394
void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*)::$_4&>(DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*)::$_4&)
./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479
std::__1::__function::__default_alloc_func<DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*)::$_4, void ()>::operator()[abi:v15000]()
./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235
void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>> const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>*)::$_4, void ()>>(std::__1::__function::__policy_storage const*)
./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716
0x11073c1a
std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda'()::operator()()
./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:0
decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda'()&>(void&&)
./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394
std::__1::__packaged_task_func<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda'(), std::__1::allocator<std::__1::function<std::__1::future<void> (std::__1::function<void ()>&&, Priority)> DB::threadPoolCallbackRunner<void, std::__1::function<void ()>>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&)::'lambda'(std::__1::function<void ()>&&, Priority)::operator()(std::__1::function<void ()>&&, Priority)::'lambda'()>, void ()>::operator()()
./build_docker/./contrib/llvm-project/libcxx/include/future:1694
0x10b362cf
std::__1::packaged_task<void ()>::operator()()
./build_docker/./contrib/llvm-project/libcxx/include/future:2069
0xc7268be
bool wide::integer<128ul, unsigned int>::_impl::operator_eq<wide::integer<128ul, unsigned int>>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
./build_docker/./base/base/../base/wide_integer_impl.h:809
bool wide::operator==<128ul, unsigned int, 128ul, unsigned int>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
./build_docker/./base/base/../base/wide_integer_impl.h:1482
StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>::operator==(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&) const
./build_docker/./base/base/../base/strong_typedef.h:42
DB::OpenTelemetry::Span::isTraceEnabled() const
./build_docker/./src/Common/OpenTelemetryTraceContext.h:65
ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>)
./build_docker/./src/Common/ThreadPool.cpp:428
0xc72a37c
ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()()
./build_docker/./src/Common/ThreadPool.cpp:0
decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(void&&)
./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394
void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&)
./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479
std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>::operator()[abi:v15000]()
./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235
void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*)
./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716
0xc728bc4
bool wide::integer<128ul, unsigned int>::_impl::operator_eq<wide::integer<128ul, unsigned int>>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
./build_docker/./base/base/../base/wide_integer_impl.h:809
bool wide::operator==<128ul, unsigned int, 128ul, unsigned int>(wide::integer<128ul, unsigned int> const&, wide::integer<128ul, unsigned int> const&)
./build_docker/./base/base/../base/wide_integer_impl.h:1482
StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>::operator==(StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag> const&) const
./build_docker/./base/base/../base/strong_typedef.h:42
DB::OpenTelemetry::Span::isTraceEnabled() const
./build_docker/./src/Common/OpenTelemetryTraceContext.h:65
ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>)
./build_docker/./src/Common/ThreadPool.cpp:428
void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const
./build_docker/./src/Common/ThreadPool.cpp:179
decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)
./build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394
void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>)
./build_docker/./contrib/llvm-project/libcxx/include/thread:284
void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*)
./build_docker/./contrib/llvm-project/libcxx/include/thread:295
0x7ffff77ab1cf
??
??:0
0x7ffff7417dd3
??
??:0
```

Probably the same as https://github.com/ClickHouse/ClickHouse/issues/55659 (closed because was happening on unsupported 23.9)

Seems related to https://github.com/ClickHouse/ClickHouse/pull/51099 (but it was merged before 23.8) and may be to https://github.com/ClickHouse/ClickHouse/pull/49474

**Comment by aadant** [comment]
```
Core was generated by `/usr/bin/clickhouse-server --config=/etc/clickhouse-server/config.xml --pid-fil'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000000012c103cf in DB::MergeTreeData::asMutableDeletingPart(std::__1::shared_ptr<DB::IMergeTreeDataPart const> const&) ()
[Current thread is 1 (Thread 0x7f3f1dc8f700 (LWP 124261))]
Missing separate debuginfos, use: yum debuginfo-install clickhouse-server-23.8.6.16-1.x86_64
(gdb) bt
#0  0x0000000012c103cf in DB::MergeTreeData::asMutableDeletingPart(std::__1::shared_ptr<DB::IMergeTreeDataPart const> const&) ()
#1  0x0000000012c9ac32 in void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::MergeTreeData::clearPartsFromFilesystemImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const> > > const&, std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >*)::$_4, void ()> >(std::__1::__function::__policy_storage const*) ()
#2  0x0000000011073c1a in std::__1::__packaged_task_func<DB::threadPoolCallbackRunner<void, std::__1::function<void ()> >(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::{lambda(std::__1::function<void ()>&&, Priority)#1}::operator()(std::__1::function<void ()>&&, Priority)::{lambda()#1}, std::__1::allocator<{lambda(std::__1::function<void ()>&&, Priority)#1}>, void ()>::operator()() ()
#3  0x0000000010b362cf in std::__1::packaged_task<void ()>::operator()() ()
#4  0x000000000c7268be in ThreadPoolImpl<ThreadFromGlobalPoolImpl<false> >::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) ()
#5  0x000000000c72a37c in void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<ThreadPoolImpl<ThreadFromGlobalPoolImpl<false> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>(ThreadPoolImpl<ThreadFromGlobalPoolImpl<false> >::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}&&)::{lambda()#1}, void ()> >(std::__1::__function::__policy_storage const*) ()
#6  0x000000000c728bc4 in void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}> >(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::{lambda()#2}>) ()
#7  0x00007ffff77ab1cf in start_thread () from /lib64/libpthread.so.0
#8  0x00007ffff7417dd3 in clone () from /lib64/libc.so.6
```

**Comment by minguyen9988** [comment]
@filimonov  which version are you seeing this on, we are seeing 23.8.6.16 official

**Comment by filimonov** [comment]
Maybe releated: https://github.com/ClickHouse/ClickHouse/commit/5db9fbed52c8c3f31b202fdd68d2d0117541d31e

**Comment by filimonov** [comment]
It seems it happens after server reach the threads limit and was not able to allotace the threads.

**Comment by filimonov** [comment]
It seems `runner` call can end with exception during the attempt of scheduling some `n+1` task - so we will leave the function and desctuct all the locals (while they can be still used by the `n` tasks which were pushed in the vector earlier. 

It looks like https://github.com/ClickHouse/ClickHouse/commit/5db9fbed52c8c3f31b202fdd68d2d0117541d31e fixes that by adding expicit wait for the tasks / threads to be finished before destruction.

BTW: maybe we can add the ability to inject the failures in task scheduling - for stress testing etc.

**Comment by Algunenano** [comment]
> BTW: maybe we can add the ability to inject the failures in task scheduling - for stress testing etc.

It was added as part of that PR -> https://github.com/ClickHouse/ClickHouse/pull/62266

**Comment by filimonov** [comment]
Fixed in 24.4.

Workaround for older versions is to avoid CANNOT_SCHEDULE_TASK / "Cannot schedule a task" exceptions by reducing the parallelism / adding monitoring to the GlobalThreadsActive metric. (expanding the limit of theads can also help, but is not recommended as it can lead to other issues / problems).

**Comment by aadant** [comment]
@filimonov is it possible to backport to 24.3 GA ?
